### PR TITLE
doc: first cut at building good containers

### PR DESCRIPTION
**Pull request by mhdawson**
First cut based on discussion in Reference
Architecture team meeting.

Signed-off-by: Michael Dawson <mdawson@devrus.com>

### CODE REVIEW

**Comment by mhdawson** [comment]
A few possibilities to discuss in the next meeting as well:

* avoiding npm start  (issues with signals being passed on)
* setting memory limits (in line with planned container max)

File: docs/development/building-good-containers.md [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+want to avoid running process as root as they may end up depending
+on capabilities that are not available when they are run `rootless`.
+
+When using docker/podman you can change the user with the `USER` command.
+Well designed base containers will have already done this, but you should
+always check what user will be used by default.
+
+```shell
+USER 1001
+RUN chown -R 1001:0 /some/directory
```
Comment (upkarlidder) [inline docs/development/building-good-containers.md:30]: I also use `--chown` flag when copying. It's a neat trick.

```
COPY --chown=1001:0 . .
RUN chmod -R g=u .
```

File: docs/development/building-good-containers.md [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+```
+
+It is recommended that you build your containers as `non-root`.
```
Comment (upkarlidder) [inline docs/development/building-good-containers.md:33]: What is meant by `building` the container? For me the `docker build` command is used to build a container. Do we instead want to say `run processes as non-root inside the container`? or something similar.

File: docs/development/building-good-containers.md [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+[multistage-build](https://docs.docker.com/develop/develop-images/multistage-build/)
+[Modern web applications on OpenShift: Part 2 -- Using chained builds](https://developers.redhat.com/blog/2018/10/23/modern-web-applications-on-openshift-part-2-using-chained-builds)
+
+It is recommended that you use multi-stage builds to minimize
+container size.
+
+
+## Add key tools for problem determination
+
+It can be difficult to add tooling/packages after the fact when
```
Comment (upkarlidder) [inline docs/development/building-good-containers.md:88]: Are we suggesting any tooling/packages in the framework? If so, we can link them here.

File: docs/development/building-good-containers.md [hunk 4]
```diff
@@ -0,0 +1,119 @@
+# Building Good Containers
+
+There are a number of recommendations based on our experience building
+Node.js applications in containers for deployments:
+
```
Comment (roastlechon) [inline docs/development/building-good-containers.md:5]: We could include a snippet about building containers with alternatives to Docker (buildah, kaniko, podman).
Reference to: https://itnext.io/docker-kaniko-buildah-209abdde5f94
https://developers.redhat.com/blog/2019/02/21/podman-and-buildah-for-docker-users#what_is_buildah_and_why_would_i_use_it_-h2

**Comment by upkarlidder** [comment]
Do we want to talk about
- sample good base image or different flavors of base node.js image
- vulnerability scanning after the image is built
- docker-ignore and keeping images lightweight

File: docs/development/building-good-containers.md (hunk 2 above)
Comment (mhdawson) [reply docs/development/building-good-containers.md:33]: +1

File: docs/development/building-good-containers.md (hunk 1 above)
Comment (mhdawson) [reply docs/development/building-good-containers.md:30]: +1

File: docs/development/building-good-containers.md (hunk 3 above)
Comment (mhdawson) [reply docs/development/building-good-containers.md:88]: I don't think we have a list right now. That would be a good addition but I think we can add in a follow on unless somebody has a good list handy.

**Comment by mhdawson** [comment]
> Do we want to talk about
> 
> * sample good base image or different flavors of base node.js image
I added a reference to the existing section in the reference architecture which covers that.
> * vulnerability scanning after the image is built
We discussed this and thought it should go into a separate security section
> * docker-ignore and keeping images lightweight
We discussed and I'll add a section to cover that.

**Comment by mhdawson** [comment]
@upkarlidder, @roastlechon I believe I've updated to address all the comments and the discussion in the last meeting.

File: docs/development/building-good-containers.md [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+* build non-root containers
+* avoid using trusted ports
```
Comment (mhamann) [inline docs/development/building-good-containers.md:7]: do you mean reserved/privileged ports here? i.e. 1-1023

Comment (jimlindeman) [reply docs/development/building-good-containers.md:7]: Yes, because of nodejs limitations with setcap (to enable non-root to use ports 1-1023).

**Review by jimlindeman — APPROVED** [review]
Looks good to me

File: docs/development/building-good-containers.md (hunk 5 above)
Comment (mhamann) [reply docs/development/building-good-containers.md:7]: right--but i think the verbiage could be more precise (e.g. privileged vs. trusted)

Comment (mhdawson) [inline docs/development/building-good-containers.md:7]:
```suggestion
* avoid reserved/privileged ports (i.e. 1-1023)
```

File: docs/development/building-good-containers.md [hunk 6]
```diff
@@ -0,0 +1,224 @@
+# Building Good Containers
+
+There are a number of recommendations based on our experience building
+Node.js applications in containers for deployments:
+
+* build non-root containers
+* avoid reserved/privileged ports (i.e. 1-1023)
```
Comment (mhdawson) [inline docs/development/building-good-containers.md:7]: @mhamann thanks for the suggestion, how does it look now?

**Pull request merged**